add new lints: `rest_pattern_accessible_field` and `unnecessary_rest_pattern`

Erk- opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/15000)*

Add a new lint that warns when using a rest pattern when destructuring a struct. This lint was requested in this comment https://github.com/rust-lang/rust-clippy/issues/10666#issuecomment-1572061535 by @xxchan.

changelog: [`rest_pattern_accessible_field`] new lint
changelog: [`unnecessary_rest_pattern`] new lint

Erk- committed:
add new lints: `rest_pattern_accessible_field` and `unnecessary_rest_pattern`

Centri3 commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,57 @@
+                pat.span,
+                "struct destructuring with rest (..)",
+                None,
+                "consider explicitly ignoring remaining fields with wildcard patterns (x: _)",
```
I'd say we should add a suggestion. It'll probably need to be converted to a late pass if you want to.

Centri3 replied:
Hmm that's unfortunate. It's probably adequate enough to shrink the `lo` portion of the pattern's span to the end of the last `PatField`. Otherwise, maybe just replace the entire pattern? It might look ok in suggestions, but it often can be finicky...

I suggest we open the FCP now and get ideas on this. It's also ok to leave this as an early lint, so long as procedural macros are ignored.

Erk- replied:
I made a bit of a hack to make something that seems to work.

Erk- replied:
After https://github.com/rust-lang/rust/pull/145783 got merged the compiler now provides the span needed here.

Centri3 commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,57 @@
+
+impl EarlyLintPass for RestWhenDestructuringStruct {
+    fn check_pat(&mut self, cx: &EarlyContext<'_>, pat: &Pat) {
+        if let PatKind::Struct(_, _, _, PatFieldsRest::Rest) = pat.kind {
```
This should check if it's from macro expansion and also call `is_from_proc_macro`, the latter requires a late pass. I see this coming up a lot in proc macro code.

Erk- replied:
I am not sure on what to call this on as `WithSearchPat` is not implemented on patterns as far as I can tell.

Centri3 replied:
Is it ok to do it on the containing expression? How would that work?

`WithSearchPat` is pretty easy to implement if you feel like it's worth doing so (I've needed this many times). Maybe there's a reason we shouldn't, but I dunno.

We can also leave a fixme, but this is quite common in proc macros so I think it's worthwhile doing it right.

Alexendoo commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,103 @@
+                        matches!(f.vis, Visibility::Public)
+                    }
+                })
+                .map(|field| field.ident(cx.tcx))
+                .filter(|pf| !fields.iter().any(|x| x.ident == *pf))
```
We can skip recovering the ident here and compare the field name against `.ident.name`

Alexendoo commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+        rustc_hir::PatKind::Missing | rustc_hir::PatKind::Err(_) => (Pat::Str(""), Pat::Str("")),
+        rustc_hir::PatKind::Wild => (Pat::Sym(kw::Underscore), Pat::Sym(kw::Underscore)),
+        rustc_hir::PatKind::Binding(binding_mode, _, ident, Some(end_pat)) => {
+            let start = match binding_mode.0 {
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Not) => Pat::Str("ref"),
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Mut) => Pat::Str("ref mut"),
+                rustc_hir::ByRef::No => {
+                    let (s, _) = ident_search_pat(ident);
+                    s
+                },
+            };
```
This can be something like

```rust
if binding_mode == BindingMode::NONE {
    ident_search_pat(ident).0
} else {
    Pat::Str(binding_mode.prefix_str())
}
```

Alexendoo commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+        },
+        rustc_hir::PatKind::Binding(binding_mode, _, ident, None) => {
+            let (s, end) = ident_search_pat(ident);
+            let start = match binding_mode.0 {
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Not) => Pat::Str("ref"),
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Mut) => Pat::Str("ref mut"),
+                rustc_hir::ByRef::No => s,
+            };
```
Same here

Alexendoo commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+            rustc_hir::RangeEnd::Excluded => (Pat::Str(".."), Pat::Str("")),
+        },
+        rustc_hir::PatKind::Range(r_start, r_end, range) => {
+            let (start, _) = match r_start {
+                Some(e) => pat_search_pat_expr_kind(e),
+                None => match range {
+                    rustc_hir::RangeEnd::Included => (Pat::Str("..="), Pat::Str("")),
+                    rustc_hir::RangeEnd::Excluded => (Pat::Str(".."), Pat::Str("")),
+                },
+            };
+
+            let (_, end) = match r_end {
+                Some(e) => pat_search_pat_expr_kind(e),
+                None => match range {
+                    rustc_hir::RangeEnd::Included => (Pat::Str(""), Pat::Str("..=")),
+                    rustc_hir::RangeEnd::Excluded => (Pat::Str(""), Pat::Str("..")),
+                },
+            };
```
The `Pat::Str("")`s here can be removed as they're immediately discarded

Alexendoo commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+    }
+}
+
+fn pat_search_pat_expr_kind(expr: &crate::PatExpr<'_>) -> (Pat, Pat) {
+    match expr.kind {
+        crate::PatExprKind::Lit { lit, negated } => {
+            let (start, end) = lit_search_pat(&lit.node);
+            if negated { (Pat::Str("!"), end) } else { (start, end) }
+        },
+        crate::PatExprKind::ConstBlock(_block) => (Pat::Str("const {"), Pat::Str("}")),
+        crate::PatExprKind::Path(path) => qpath_search_pat(&path),
+    }
+}
```
You can add imports for `rustc_hir::{PatKind, PatExpr, PatExprKind}`. `Pat` I understand conflicts so is fine to leave prefixed

Alexendoo commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,103 @@
+                .fields
+                .iter()
+                .filter(|f| {
+                    if a.did().is_local() {
+                        true
+                    } else {
+                        matches!(f.vis, Visibility::Public)
+                    }
```
This case would also require the `..` to be kept

ada4a reviewed (commented):
Having the messages be a bit more specific really goes a long way!

Some final touches, and then it should be good to go^^

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/30c73fef627d359eddeac7e422ebc837afb869cf..2caeede62bcea3ed6a65514c3a6383faae9ff563)*

Erk- commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -541,6 +541,99 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+            let (_, end) = pat_search_pat(tcx, p);
+            (Pat::Str("box"), end)
+        },
+        PatKind::Deref(_) => (Pat::Str("deref!("), Pat::Str(")")),
```
The documentation for the nightly feature only made references to the `deref!(...)` version, but it seems like all the forms are accepted. So I have changed it in https://github.com/rust-lang/rust-clippy/pull/15000/commits/e95f1417f00a86d0aa7a0aab0aa5b0fa1e7c2559.

cf: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=a8f3af47ee51b6b0345619c34e4a653c

Erk- commented:
@Alexendoo Do you have some time to have a second look at this pr?

Jarcho commented:
We do one lint per test file so you'll want have to split the test.

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,204 @@
+    fn check_pat(&mut self, cx: &rustc_lint::LateContext<'tcx>, pat: &'tcx rustc_hir::Pat<'tcx>) {
+        if let rustc_hir::PatKind::Struct(path, fields, Some(dotdot)) = pat.kind
+            && !pat.span.in_external_macro(cx.tcx.sess.source_map())
+            && !is_from_proc_macro(cx, pat)
```
This should be the last thing you check. It's both slow and doesn't filter out much.

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,204 @@
+                .opt_def_id()
+                .map_or(VariantIdx::ZERO, |x| a.variant_index_with_id(x));
+
+            if a.variants()[vid].fields.is_empty() {
```
Please index `variants` only once.

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,204 @@
+        {
+            let vid = qty
+                .opt_def_id()
+                .map_or(VariantIdx::ZERO, |x| a.variant_index_with_id(x));
```
Defaulting the variant index is wrong. This can't fail without the compiler screwing up, but it should default to not linting.

Erk- replied:
I have put it as a guard instead

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,204 @@
+            if fmt_fields.is_empty() {
+                // The struct either has no unnamed member or only has unnameable members left.
+                return;
+            }
```
For this I would find the index of the first unnamed field and then do the formatting.

```rust
let module = cx.tcx.parent_module(pat.hir_id);
let f = |f| fields.iter.all(|x| x.ident.name != f.name)) && f.vis.is_accessible_from(module, cx.tcx);

let Some(first) = variant.fields.iter().position(f) else { return; };

let mut fmt_fields = iter::once(variant.fields[first])
  .chain(variant.fields[first + 1..].iter()
    .filter(f))
    .format_with(", ", |f, write| write(format_args!("{}: _", f.name.as_str())))
  );
```

Erk- replied:
The types used there makes it a bit annoying to do, but I think I did something that works.

Jarcho reviewed (changes requested):
This should be a single lint pass since they do basically all the same work. It looks like there was some confusion here since `RestPatternAccessibleField ` has some of the messages for the other lint pass's lint.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/68d2b73340bb762e9e57ee4e009bec2c13d5a422..a81b313941638c0e1921d8cd0b32d3c6cc48cb01)*

Erk- commented:
@rustbot ready

I think the combined pass makes it a bit harder to follow what is happening because they are quite intertwined, but it is okay still.

Jarcho commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -588,6 +589,98 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+        },
+        PatKind::TupleStruct(path, _, _) => {
+            let (start, _) = qpath_search_pat(&path);
+            (start, Pat::Str(")"))
```
Parenthesis are stripped before comparing the text.

Jarcho commented on clippy_utils/src/check_proc_macro.rs:
```diff
@@ -588,6 +589,98 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
+            (start, end)
+        },
+        PatKind::Never => (Pat::Str("!"), Pat::Str("")),
+        PatKind::Tuple(_, _) => (Pat::Str("("), Pat::Str(")")),
```
Parenthesis are stripped before comparing the text.

Jarcho reviewed (changes requested):
You're trying too hard to avoid writing an explicit loop. It makes things a lot simpler when you do it all in one loop:
```rust
let mut missing_sugg = String::new();
let mut needs_dd = false;
for f in variant.fields {
    if f.vis.is_accessible_from(cx.tcx.parent_module(pat.hir_id), cx.tcx) {
        if !fields.iter().any(|x| x.ident.name == f.name) {
            if !missing_sugg.is_empty() {
                missing_sugg.push_str(", ");
            }
            missing_sugg.push_str(f.name.as_str());
        }
    } else {
        needs_dd = true;
    }
}

if (missing_sugg.is_empty() && !needs_dd)
    || pat.span.in_external_macro(cx.tcx.sess.source_map())
    || is_from_proc_macro(cx, pat)
{
    return;
}

if !missing_sugg.is_empty() {
    let sugg_sp = if needs_dd {
        missing_sugg.push_str(", ");
        dotdot.shrink_to_lo();
    } else {
        dotdot
    };
    
    lint()
}

if !needs_dd {
    lint()
}
```

Note you want the macro checks to run later since they're both slower and don't filter out many cases.

---


Neither of the tuple patterns can show up in the proc-macro check so I'm fine if you want to just use `""` and leave a note. I have a separate PR that completely changes how this code works so it will be fixed eventually.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

Jarcho commented:
Restarting the FCP since this has two lints now and we have a bot supported thing now.

@rustbot label lint-nominated

r? Jarcho

Jarcho commented:
You'll also need to keep the `..` if the type has the `#[non_exhaustive]` attribute. You have `VariantDef::is_field_list_non_exhaustive` for that.

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+    #[clippy::version = "1.94.0"]
+    pub UNNECESSARY_REST_PATTERN,
+    restriction,
+    "unnecessary rest (..) in destructuring expression"
```
```suggestion
    "unnecessary rest pattern (`..`) in destructuring expression"
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
````diff
@@ -0,0 +1,201 @@
+    ///
+    /// let S { a, b, c } = s;
+    /// ```
+    #[clippy::version = "1.94.0"]
````
```suggestion
    #[clippy::version = "1.98.0"]
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
````diff
@@ -0,0 +1,201 @@
+    ///
+    /// let S { a, b, c: _ } = s;
+    /// ```
+    #[clippy::version = "1.94.0"]
````
```suggestion
    #[clippy::version = "1.98.0"]
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Disallow the use of rest patterns that are unnecessary.
```
This isn't descriptive enough imo. How about:
```suggestion
    /// Disallow the use of rest patterns that don't capture any fields.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+    /// ### What it does
+    /// Disallow the use of rest patterns that are unnecessary.
+    ///
+    /// ### Why is this bad?
```
For restriction lints, this heading should be different:
```suggestion
    /// ### Why restrict this?
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+    /// ### What it does
+    /// Disallow the use of rest patterns for accesible fields
+    ///
+    /// ### Why is this bad?
```
ditto
```suggestion
    /// ### Why restrict this?
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+    #[clippy::version = "1.94.0"]
+    pub REST_PATTERN_ACCESSIBLE_FIELD,
+    restriction,
+    "rest (..) used for accessible field"
```
```suggestion
    "rest pattern (`..`) used for accessible field"
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+                            dotdot,
+                            message,
+                            fmt_fields,
+                            rustc_errors::Applicability::MachineApplicable,
```
ditto
```suggestion
                            Applicability::MachineApplicable,
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+                            dotdot,
+                            message,
+                            "",
+                            rustc_errors::Applicability::MachineApplicable,
```
ditto
```suggestion
                            Applicability::MachineApplicable,
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

ada4a commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,201 @@
+                    "consider explicitly ignoring fields with wildcard patterns (`x: _`)"
+                } else {
+                    "consider explicitly ignoring remaining fields with wildcard patterns (`x: _`)"
+                };
```
nit: all of theses can be done inside `span_lint_and_then`. Though this might no longer be relevant after Jason's comment

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/e74196a266eb57f2960e68e0eccc558600c8bae6..021bab52afac1029cb88be723b955094dc89e30c)*

Erk- commented:
@rustbot ready

The explicit loop did make it quite a bit nicer.

> You'll also need to keep the `..` if the type has the `#[non_exhaustive]` attribute. You have `VariantDef::is_field_list_non_exhaustive` for that.

I ended up using `VariantDef::field_list_has_applicable_non_exhaustive` since otherwise a local struct with `#[non_exhaustive]` would be handled incorrectly.

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,171 @@
+                || is_from_proc_macro(cx, pat)
+            {
+                return;
+            }
```
It looks like you were going if "lints and in macro", but you have the precedence wrong (`&&` is higher). You get the correct answer, but you'll run `is_from_proc_macro` even when there's nothing to emit.

I would just write this as:

```rust
if (missing_suggestions.is_empty() && needs_dotdot)
  || pat.span.in_external_macro(cx.tcx.sess.source_map())
  || is_from_proc_macro(cx, pat)
{
  return;
}
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/3be3a9cf09fbae8b9f8c359c8df3acef980f7ec9..21edfe2ca450e36b3cdccbfe5b97eab7d815619b)*

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,171 @@
+                );
+            }
+
+            if missing_suggestions.is_empty() && !needs_dotdot {
```
`else if !needs_dotdot`

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/3be3a9cf09fbae8b9f8c359c8df3acef980f7ec9..21edfe2ca450e36b3cdccbfe5b97eab7d815619b)*

Erk- commented:
@rustbot ready

For some reason GitHub did not give me any notifications for this, but the review should be resolved now.

Jarcho commented on clippy_lints/src/rest_when_destructuring_struct.rs:
```diff
@@ -0,0 +1,169 @@
+        {
+            // dbg!(&variant, &fields);
+            let mut missing_suggestions = String::new();
+            let mut needs_dotdot = false;
```
Minor, but this can be initialized with `variant.field_list_has_applicable_non_exhaustive()`

ada4a replied:
You probably also want the delete the commented-out debugprint^^

Jarcho reviewed (approved):
One minor nit that isn't really important. Can you please also squash the commits down.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15000/d22d01ee1332567e7bafd610f1aa9115e10fc0cb..fa7fe29db0fab8e1bb2a6492529a89bb9b0ddb78)*

Erk- commented:
Resolved the last two small nits.

Also updated the rustc version.

I squashed the commits down to a single commit, for future reference I did leave a branch on my fork with the full list of commits: https://github.com/Erk-/rust-clippy/tree/all-the-commits.

Jarcho commented:
Thank you.

@rustbot merge

Jarcho commented:
Ping @ada4a since you have a review blocking this. I assume everything was done?
